crates/chain/src/tx_graph.rs

```diff
@@ -232,176 +498,446 @@ impl TxGraph {
         let mut additions = Additions::default();
 
-        for (&txid, update_tx) in &update.txs {
-            if self.get_tx(txid).is_some() {
-                continue;
-            }
-
-            match update_tx {
-                TxNode::Whole(tx) => {
-                    if matches!(self.txs.get(&txid), None | Some(TxNode::Partial(_))) {
-                        additions.tx.insert(tx.clone());
-                    }
+        for (&txid, (update_tx_node, _, update_last_seen)) in &update.txs {
+            let prev_last_seen: u64 = match (self.txs.get(&txid), update_tx_node) {
+                (None, TxNodeInternal::Whole(update_tx)) => {
+                    additions.tx.insert(update_tx.clone());
+                    0
                 }
-                TxNode::Partial(partial) => {
-                    for (&vout, update_txout) in partial {
-                        let outpoint = OutPoint::new(txid, vout);
-
-                        if self.get_txout(outpoint) != Some(update_txout) {
-                            additions.txout.insert(outpoint, update_txout.clone());
-                        }
-                    }
+                (None, TxNodeInternal::Partial(update_txos)) => {
+                    additions.txout.extend(
+                        update_txos
+                            .iter()
+                            .map(|(&vout, txo)| (OutPoint::new(txid, vout), txo.clone())),
+                    );
+                    0
+                }
+                (Some((TxNodeInternal::Whole(_), _, last_seen)), _) => *last_seen,
+                (
+                    Some((TxNodeInternal::Partial(_), _, last_seen)),
+                    TxNodeInternal::Whole(update_tx),
+                ) => {
+                    additions.tx.insert(update_tx.clone());
+                    *last_seen
+                }
+                (
+                    Some((TxNodeInternal::Partial(txos), _, last_seen)),
+                    TxNodeInternal::Partial(update_txos),
+                ) => {
+                    additions.txout.extend(
+                        update_txos
+                            .iter()
+                            .filter(|(vout, _)| !txos.contains_key(*vout))
+                            .map(|(&vout, txo)| (OutPoint::new(txid, vout), txo.clone())),
+                    );
+                    *last_seen
                 }
+            };
+
+            if *update_last_seen > prev_last_seen {
+                additions.last_seen.insert(txid, *update_last_seen);
             }
         }
 
+        additions.anchors = update.anchors.difference(&self.anchors).cloned().collect();
+
         additions
     }
+}
 
-    /// Returns the resultant [`Additions`] if the given transaction is inserted. Does not actually
-    /// mutate [`Self`].
+impl<A: Anchor> TxGraph<A> {
+    /// Get all heights that are relevant to the graph.
+    pub fn relevant_heights(&self) -> impl Iterator<Item = u32> + '_ {
```

Comment:
This API has been added but isn't used anywhere and not tested.

The implementation here does not need a `HashSet`. The set is ordered by height anyway right? Just remove items that have the same height as the last one.

Reply:
This was intended for obtaining blocks of given heights for the `LocalChain`. The `HashSet` is so we don't have duplicate heights, not for order.

Reply:
I understand. you can deduplicate them without a hashset here. Duplicate ones will be right after each other because it's in order.

Reply:
Fixed.